### PR TITLE
[API] Add notification endpoint

### LINKED ISSUE #4811: Export notifications through the API
- Gitea version (or commit ref): c784ac5
- Git version: 2.17.1
- Operating system: Fedora 27
- Database (use `[x]`):
  - [X] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [X] Yes (provide example URL)
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description

It would be nice to access notifications via the API, just like on GitHub: https://api.github.com/notifications
...

### PR DESCRIPTION
**Pull request by 6543**
close #4811

This is the main PR adding Notification Endpoint ...

## Changes
* [x] Basic api Notification Struct
* [x] Endpoints
  - [x] `GET /notifications` [[spec](https://developer.github.com/v3/activity/notifications/#list-your-notifications)]
  - [x] `PUT /notifications` [[spec](https://developer.github.com/v3/activity/notifications/#mark-as-read)]
  - [x] `GET /repos/{owner}/{repo}/notifications` [[spec](https://developer.github.com/v3/activity/notifications/#list-your-notifications-in-a-repository)]
  - [x] `PUT /repos/{owner}/{repo}/notifications` [[spec](https://developer.github.com/v3/activity/notifications/#mark-notifications-as-read-in-a-repository)]
  - [x] `GET /notifications/threads/{id}` [[spec](https://developer.github.com/v3/activity/notifications/#view-a-single-thread)]
  - [x] `PATCH /notifications/threads/{id}` [[spec](https://developer.github.com/v3/activity/notifications/#mark-a-thread-as-read)]
* [x] latest_comment_url
  -  [x]  add endpoint for single comment (#9486) (#9494)
  -  [x]  add Issue.LatestComment
* [x] add Test
## future imprufements:
* optional add filter for pinned
* implement NotificationReason (ea17f03087838d0663be067e79380d04ddbf2165) no info in DB jet
*  `PATCH /notifications/threads/{id}` option to mark as pin/unread
* endpoint for mobile clients -> #9595

### CODE REVIEW

File: models/notification.go [hunk 1]
```diff
@@ -238,6 +239,42 @@ func notificationsForUser(e Engine, user *User, statuses []NotificationStatus, p
 	return
 }
 
+// APIFormat converts a Notification to api.NotificationThread
+func (n *Notification) APIFormat() *api.NotificationThread {
```
Comment (lunny) [inline models/notification.go:298]: I would like to move all the convertion to `route/conversion` package

Comment (6543) [reply models/notification.go:298]: @lunny should I creat a refactor-PR for all other APIFormat() func?

**Review by zeripath — APPROVED** [review]

**Comment by 6543** [comment]
@zeripath nice for the lgtm but i think this is to early! maby you mixed up #9494 ?

**Comment by zeripath** [comment]
I didn't notice that you hadn't finished! that's what I get for reviewing whilst drinking!

What you've written so far looks reasonable!

🎄 ☃️ Merry Christmas ☃️ 🎄

File: models/notification.go (hunk 1 above)
Comment (6543) [reply models/notification.go:298]: cant find `route/conversion`

File: routers/api/v1/notify/repo.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
+// Copyright 2016 The Gitea Authors. All rights reserved.
```
Comment (6543) [inline routers/api/v1/notify/repo.go]:
```suggestion
// Copyright 2020 The Gitea Authors. All rights reserved.
```

File: routers/api/v1/notify/threads.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
+// Copyright 2016 The Gitea Authors. All rights reserved.
```
Comment (6543) [inline routers/api/v1/notify/threads.go]:
```suggestion
// Copyright 2020 The Gitea Authors. All rights reserved.
```

File: routers/api/v1/notify/user.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
+// Copyright 2016 The Gitea Authors. All rights reserved.
```
Comment (6543) [inline routers/api/v1/notify/user.go]:
```suggestion
// Copyright 2020 The Gitea Authors. All rights reserved.
```

File: integrations/api_notification_test.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+// license that can be found in the LICENSE file.
+
+package integrations
+
+import (
+	"code.gitea.io/gitea/models"
+	api "code.gitea.io/gitea/modules/structs"
+	"fmt"
+	"net/http"
+	"testing"
```
Comment (adelowo) [inline integrations/api_notification_test.go:10]: Proper formatting.

**Review by adelowo — CHANGES_REQUESTED** [review]

File: integrations/api_notification_test.go (hunk 5 above)
Comment (6543) [reply integrations/api_notification_test.go:10]: and there it happend ... :sweat_smile:

Comment (6543) [reply integrations/api_notification_test.go:10]: @adelowo done (sometimes my IDE mess up with this)

File: integrations/api_notification_test.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
```
Comment (6543) [inline integrations/api_notification_test.go]:
```suggestion
// Copyright 2020 The Gitea Authors. All rights reserved.
```

**Comment by 6543** [comment]
@adelowo done
PS: this PR include to PR (see commit & commit.messages) because it depend on them :(
hope they get merged first ...

**Comment by 6543** [comment]
@adelowo done

**Comment by 6543** [comment]
when #9373 is merged ... CI should pass :D

File: models/notification.go [hunk 7]
```diff
@@ -238,21 +294,125 @@ func notificationsForUser(e Engine, user *User, statuses []NotificationStatus, p
...
+	return nil
+}
+
 // GetRepo returns the repo of the notification
 func (n *Notification) GetRepo() (*Repository, error) {
-	n.Repository = new(Repository)
-	_, err := x.
-		Where("id = ?", n.RepoID).
-		Get(n.Repository)
+	err := n.loadRepo(x)
```
Comment (lafriks) [inline models/notification.go]: Just move to return, no need for err variable:
```go
return n.Repository, n.loadRepo(x)
```

File: models/notification.go [hunk 8]
```diff
@@ -238,21 +294,125 @@ func notificationsForUser(e Engine, user *User, statuses []NotificationStatus, p
...
 	return n.Repository, err
 }
 
 // GetIssue returns the issue of the notification
 func (n *Notification) GetIssue() (*Issue, error) {
-	n.Issue = new(Issue)
-	_, err := x.
-		Where("id = ?", n.IssueID).
-		Get(n.Issue)
+	err := n.loadIssue(x)
```
Comment (lafriks) [inline models/notification.go]: Same

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/notification.go (hunk 7 above)
Comment (6543) [inline models/notification.go]:
```suggestion
```

File: models/notification.go [hunk 9]
```diff
@@ -238,21 +294,125 @@ func notificationsForUser(e Engine, user *User, statuses []NotificationStatus, p
...
+}
+
 // GetRepo returns the repo of the notification
 func (n *Notification) GetRepo() (*Repository, error) {
-	n.Repository = new(Repository)
-	_, err := x.
-		Where("id = ?", n.RepoID).
-		Get(n.Repository)
+	err := n.loadRepo(x)
 	return n.Repository, err
```
Comment (6543) [inline models/notification.go]:
```suggestion
	return n.Repository, n.loadRepo(x)
```

File: models/notification.go [hunk 10]
```diff
@@ -238,21 +294,124 @@ func notificationsForUser(e Engine, user *User, statuses []NotificationStatus, p
...
+	return n.Repository, n.loadRepo(x)
 }
 
 // GetIssue returns the issue of the notification
 func (n *Notification) GetIssue() (*Issue, error) {
-	n.Issue = new(Issue)
-	_, err := x.
-		Where("id = ?", n.IssueID).
-		Get(n.Issue)
+	err := n.loadIssue(x)
```
Comment (6543) [inline models/notification.go]:
```suggestion
```

File: models/notification.go [hunk 11]
```diff
@@ -238,21 +294,124 @@ func notificationsForUser(e Engine, user *User, statuses []NotificationStatus, p
...
 	return n.Issue, err
```
Comment (6543) [inline models/notification.go]:
```suggestion
	return n.Issue, n.loadIssue(x)
```

**Review by lafriks — APPROVED** [review]

**Pull request merged**